core/array/sum_spec.rb

```diff
@@ -0,0 +1,49 @@
+require File.expand_path('../../../spec_helper', __FILE__)
+require File.expand_path('../fixtures/sum', __FILE__)
+
+ruby_version_is '2.4' do
+  describe "Array#sum" do
+    it "returns the sum of elements" do
+      [1, 2, 3].sum.should == 6
+    end
+
+    it "applies a block to each element before adding if it's given" do
+      [1, 2, 3].sum { |i| i * 10 }.should == 60
+    end
+
+    it "returns init value if array is empty" do
+      [].sum(-1).should == -1
+    end
+
+    it "returns 0 if array is empty and init is omited" do
+      [].sum.should == 0
+    end
+
+    it "adds init value to the sum of elemens" do
+      [1, 2, 3].sum(10).should == 16
+    end
+
+    it "can be used for non-numeric objects by providing init value" do
+      ["a", "b", "c"].sum("").should == "abc"
+    end
+
+    it 'raises TypeError if any element are not numeric' do
+      lambda { ["a"].sum }.should raise_error(TypeError)
+    end
+
+    it 'raises TypeError if any element cannot be added to init value' do
+      lambda { [1].sum([]) }.should raise_error(TypeError)
+    end
+
+    it "may not respect redefinishion of :+ method of core classes" do
+      lambda {
+        ArraySpec::AdditionIsOverriden.call_explicit_adding
+      }.should raise_error(RuntimeError, "Redefined method is called")
+
+
+      lambda {
+        ArraySpec::AdditionIsOverriden.call_sum
+      }.should_not raise_error
+    end
```

Comment:
I think testing refinements here is a bit out of scope.
More importantly, I want Ruby implementations without refinements to be able to run the rest of the specs.

What we could test and might be interesting would be to monkey patch `+`, and run that with `ruby_exe`.
But we see that MRI does not respect the monkey patch here :(
```ruby
class Integer; def +(o); 42; end; end; p [1,2,3].sum # => 6
```

Maybe what we want to test here is that `+` is called on other objects.
A `mock()` could be used for that.

Reply:
@eregon I see. I've updated test case

Reply:
I am sorry I was not clear. I meant testing monkey patching here is not interesting since MRI doesn't respect it. And testing that it doesn't crash is not very interesting.
I'll replace this test with using a mock() to ensure `+` is called as expected in the general case.

Reply:
@eregon this behaviour is mentioned in the [documentation](https://ruby-doc.org/core-2.4.0/Array.html#method-i-sum):

> #sum method may not respect method redefinition of “+” methods such as Integer#+.

So I believe it should be a part of specs

Reply:
I see, did not know that.
I think the purpose is mostly for Float to have an accurate sum compensating for errors (see [the original Feature in MRI's bug tracker](https://bugs.ruby-lang.org/issues/12217)). That would be an additional thing to test that is very relevant.

Integer being treated specially seems just a low-level optimization by MRI that I don't think make sense to force on other implementations, that's why I am rather against it. The `may` expresses this might not be the case and it's implementation-specific behavior (so should be part of the implementation-specific tests)